From 4b1d2f1ce176d6c21ed883b961c42af5537fd877 Mon Sep 17 00:00:00 2001
From: Jakub Filak <jfilak@redhat.com>
Date: Wed, 14 May 2014 10:22:13 +0200
Subject: [LIBREPORT PATCH] Make make_description() output less confusing

Implement ideas of Martin Milata:

I'd remove the empty line before the "Bugzilla:" line, it might not be
immediately clear to which problem the line belongs if there's several
problems in the list.

Also, what about something like:

  Reported:       https://bugzilla.redhat.com/1234
                  https://retrace.fedoraproject.org/report/6789

http://post-office.corp.redhat.com/archives/abrt-devel-list/2013-December/msg00171.html

Related to rhbz#1036585

Signed-off-by: Jakub Filak <jfilak@redhat.com>

mmilata: comment nitpicking

Conflicts:

	po/POTFILES.in
	src/lib/make_descr.c
	tests/make_description.at
---
 po/POTFILES.in            |  2 ++
 src/lib/make_descr.c      | 22 +++++++++++++---------
 tests/make_description.at | 35 +++++++++++++++--------------------
 3 files changed, 30 insertions(+), 29 deletions(-)

diff --git a/po/POTFILES.in b/po/POTFILES.in
index 61c9c85..70415e8 100644
--- a/po/POTFILES.in
+++ b/po/POTFILES.in
@@ -12,6 +12,8 @@ src/gui-wizard-gtk/wizard.glade
 src/lib/abrt_sock.c
 src/lib/create_dump_dir.c
 src/lib/event_config.c
+src/lib/json.c
+src/lib/make_descr.c
 src/lib/parse_options.c
 src/lib/abrt_curl.c
 src/lib/client.c
diff --git a/src/lib/make_descr.c b/src/lib/make_descr.c
index e4ae6d5..12f6ac5 100644
--- a/src/lib/make_descr.c
+++ b/src/lib/make_descr.c
@@ -100,7 +100,6 @@ char *make_description(problem_data_t *problem_data, char **names_to_skip,
         }
     }
 
-    bool append_empty_line = !empty;
     if (desc_flags & MAKEDESC_SHOW_URLS)
     {
         const char *reported_to = get_problem_item_content_or_NULL(problem_data, FILENAME_REPORTED_TO);
@@ -108,6 +107,11 @@ char *make_description(problem_data_t *problem_data, char **names_to_skip,
         {
             GList *reports = read_entire_reported_to_data(reported_to);
 
+            /* The value part begins on 17th column */
+            /*                        0123456789ABCDEF*/
+            const char *pad_prefix = "                ";
+            char *first_prefix = xasprintf("%s%*s", _("Reported:"), 16 - strlen(_("Reported:")), "");
+            const char *prefix     = first_prefix;
             for (GList *iter = reports; iter != NULL; iter = g_list_next(iter))
             {
                 const report_result_t *const report = (report_result_t *)iter->data;
@@ -115,21 +119,21 @@ char *make_description(problem_data_t *problem_data, char **names_to_skip,
                 if (report->url == NULL)
                     continue;
 
-                if (append_empty_line)
-                    strbuf_append_char(buf_dsc, '\n');
-                append_empty_line = false;
-                empty = false;
+                strbuf_append_strf(buf_dsc, "%s%s\n", prefix, report->url);
 
-                int pad = 16 - (strlen(report->label) + 2);
-                if (pad < 0) pad = 0;
-                strbuf_append_strf(buf_dsc, "%s: %*s%s\n", report->label, pad, "", report->url);
+                if (prefix == first_prefix)
+                {   /* Only the first URL is prefixed by 'Reported:' */
+                    empty = false;
+                    prefix = pad_prefix;
+                }
             }
 
+            free(first_prefix);
             list_free_with_free(reports);
         }
     }
 
-    append_empty_line = !empty;
+    bool append_empty_line = !empty;
     if (desc_flags & MAKEDESC_SHOW_FILES)
     {
         /* Print file info. Format:
diff --git a/tests/make_description.at b/tests/make_description.at
index 8e337e0..46b63e0 100644
--- a/tests/make_description.at
+++ b/tests/make_description.at
@@ -28,7 +28,7 @@ int main(int argc, char **argv)
 
     description = make_description(pd, /*skipped names*/NULL, CD_MAX_TEXT_SIZE, MAKEDESC_SHOW_URLS);
     char *expected = xasprintf("%s: %*s%s\n",
-            "Bugzilla", 14 - strlen("Bugzilla"), "", "https://bugzilla.redhat.com/1000000");
+            "Reported", 14 - strlen("Reported"), "", "https://bugzilla.redhat.com/1000000");
 
     if (strcmp(expected, description) != 0)
     {
@@ -50,10 +50,11 @@ int main(int argc, char **argv)
             "ABRT Server: URL=https://bug-report.itos.redhat.com\n");
 
     description = make_description(pd, /*skipped names*/NULL, CD_MAX_TEXT_SIZE, MAKEDESC_SHOW_URLS);
-    expected = xasprintf("%s: %*s%s\n%s: %*s%s\n%s: %*s%s\n",
-            "Bugzilla",    14 - strlen("Bugzilla"),    "", "https://bugzilla.redhat.com/1000000",
-            "RHTSupport",  14 - strlen("RHTSupport"),  "", "https://access.redhat.com/home",
-            "ABRT Server", 14 - strlen("ABRT Server"), "", "https://bug-report.itos.redhat.com");
+    expected = xasprintf(
+           /*0123456789ABCDEF*/
+            "Reported:       https://bugzilla.redhat.com/1000000\n"
+            "                https://access.redhat.com/home\n"
+            "                https://bug-report.itos.redhat.com\n");
 
     if (strcmp(expected, description) != 0)
     {
@@ -98,16 +99,13 @@ int main(int argc, char **argv)
             "%s: %*s%s\n"
             "%s: %*s%s\n"
             "%s: %*s%s\n"
-            "\n"
-            "%s: %*s%s\n"
-            "%s: %*s%s\n"
-            "%s: %*s%s\n",
+           /*0123456789ABCDEF*/
+            "Reported:       https://bugzilla.redhat.com/1000000\n"
+            "                https://access.redhat.com/home\n"
+            "                https://bug-report.itos.redhat.com\n",
             FILENAME_COUNT,      14 - strlen(FILENAME_COUNT),      "", "0",
             FILENAME_EXECUTABLE, 14 - strlen(FILENAME_EXECUTABLE), "", "/usr/bin/sh",
-            FILENAME_PACKAGE,    14 - strlen(FILENAME_PACKAGE),    "", "libreport",
-            "Bugzilla",          14 - strlen("Bugzilla"),          "", "https://bugzilla.redhat.com/1000000",
-            "RHTSupport",        14 - strlen("RHTSupport"),        "", "https://access.redhat.com/home",
-            "ABRT Server",       14 - strlen("ABRT Server"),       "", "https://bug-report.itos.redhat.com");
+            FILENAME_PACKAGE,    14 - strlen(FILENAME_PACKAGE),    "", "libreport");
 
     if (strcmp(expected, description) != 0)
     {
@@ -127,19 +125,16 @@ int main(int argc, char **argv)
             "%s: %*s%s\n"
             "%s: %*s%s\n"
             "%s: %*s%s\n"
-            "\n"
-            "%s: %*s%s\n"
-            "%s: %*s%s\n"
-            "%s: %*s%s\n"
+           /*0123456789ABCDEF*/
+            "Reported:       https://bugzilla.redhat.com/1000000\n"
+            "                https://access.redhat.com/home\n"
+            "                https://bug-report.itos.redhat.com\n"
             "\n"
             "%s: %*sText file, %llu bytes\n"
             "%s: %*sText file, %llu bytes\n",
             FILENAME_COUNT,      14 - strlen(FILENAME_COUNT),      "", "0",
             FILENAME_EXECUTABLE, 14 - strlen(FILENAME_EXECUTABLE), "", "/usr/bin/sh",
             FILENAME_PACKAGE,    14 - strlen(FILENAME_PACKAGE),    "", "libreport",
-            "Bugzilla",          14 - strlen("Bugzilla"),          "", "https://bugzilla.redhat.com/1000000",
-            "RHTSupport",        14 - strlen("RHTSupport"),        "", "https://access.redhat.com/home",
-            "ABRT Server",       14 - strlen("ABRT Server"),       "", "https://bug-report.itos.redhat.com",
             FILENAME_BACKTRACE,  14 - strlen(FILENAME_BACKTRACE),  "", (long long unsigned)strlen(backtrace),
             FILENAME_REPORTED_TO,14 - strlen(FILENAME_REPORTED_TO),"", (long long unsigned)strlen(get_problem_item_content_or_NULL(pd, FILENAME_REPORTED_TO))
             );
-- 
1.8.3.1

